### PR TITLE
feat(tests): Unit tests for all queries/mutations

### PR DESCRIPTION
**Pull request by zarghamkhandev**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/query/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the new behavior?

Partially implements #23 

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/ng-query/src/lib/tests/query.spec.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    expect(result).toMatchObject({
+      status: 'success',
+      data: 'result2',
+      isLoading: false,
+      isFetching: false,
+      isFetched: true,
+      isSuccess: true,
+    });
+
+    /* expect(query.result$).toBeObservable( */
```
Comment (zarghamkhandev) [inline packages/ng-query/src/lib/tests/query.spec.ts:54]: @NetanelBasal  I wanted to have marble tests in here using `jest-marbles`, but `query.result$` emits only 1 frame. Could you have a look please ?

Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/query.spec.ts:54]: I'd use https://github.com/hirezio/observer-spy

File: packages/ng-query/src/lib/tests/query.spec.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      isFetching: false,
+      isFetched: true,
+      isError: true,
+      failureCount: 1,
+      failureReason: { message: 'Some error' },
+    });
+  });
+
+  it.skip('should call onSuccess callback', async () => {
+    const onSuccess = jest.fn();
```
Comment (zarghamkhandev) [inline packages/ng-query/src/lib/tests/query.spec.ts:113]: @NetanelBasal  `onSuccess`, `onError`, `onSettled` callbacks are never called. Do we want to fix it or omit them from options ?

Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/query.spec.ts:113]: Fix it. I'm wondering why it's not working.

Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:113]: tried to debug but no luck, will create a separate issue.

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 1 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:54]: let me try it.

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 2 above)
Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/query.spec.ts:113]: Have you tried it in the playground?

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 1 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts]: oh this works like a charm. 🔥 marble tests are hard.

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 2 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:118]: just tried in playground, the callback works there, not in tests, will investigate further.

File: packages/ng-query/src/lib/tests/query.spec.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  let client: QueryCore;
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({});
+    useQuery = TestBed.inject(QueryProvider);
+    client = TestBed.inject(QueryClient);
+  });
+
+  it('should properly execute query', () => {
+    useQuery(['key0'], simpleFetcher, { staleTime: 1000 });
```
Comment (NetanelBasal) [inline packages/ng-query/src/lib/tests/query.spec.ts:23]: What's the difference between simpleFetcher and simpleFetcher?

File: packages/ng-query/src/lib/tests/query.spec.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+      failureCount: 1,
+      failureReason: { message: 'Some error' },
+    });
+  });
+
+  it.skip('should call onSuccess callback', async () => {
+    const onSuccess = jest.fn();
+
+    useQuery(['key6'], simpleFetcher, {
+      onSuccess: () => {
```
Comment (NetanelBasal) [inline packages/ng-query/src/lib/tests/query.spec.ts:128]: You're not using the spy. It should be:

```ts
onSuccess,
```

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 3 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:23]: you mean `simpleFetcher` and `fetcher` ? they are utils I copied from vue-query tests

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 4 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:128]: here I was testing if it gets logged, but it does not. I am having no idea why its not being called in tests but it does in playground.

File: packages/ng-query/src/lib/tests/mutation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+      isLoading: true,
+      isError: false,
+      isSuccess: false,
+      data: undefined,
+      error: null,
+    });
+  });
+
+  it('should return error when request fails', async () => {
+    const mutation = useMutation(errorMutator);
```
Comment (zarghamkhandev) [inline packages/ng-query/src/lib/tests/mutation.spec.ts:50]: if `mutationFn` throws error, this error is not caught and gets to zonejs. Tested in playground, is same there as well. @NetanelBasal

Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/mutation.spec.ts:50]: >  this error is not caught and gets to zonejs

Can you elaborate more on this? What do you mean?

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 3 above)
Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/query.spec.ts:23]: But simpleFetcher isn't a function that returns an observable, so it's irrelevant.

File: packages/ng-query/src/lib/tests/mutation.spec.ts (hunk 5 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/mutation.spec.ts:50]: my bad. Actually, the expected behavior is correct. Never mind. The only problem I'm left with this test case is that I cannot get rid of this log message even though the test case is successful. 

image

Comment (NetanelBasal) [reply packages/ng-query/src/lib/tests/mutation.spec.ts:50]: https://github.com/facebook/jest/issues/5785

Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/mutation.spec.ts:53]: thank you, that worked pretty well

File: packages/ng-query/src/lib/tests/query.spec.ts (hunk 3 above)
Comment (zarghamkhandev) [reply packages/ng-query/src/lib/tests/query.spec.ts:23]: I updated the functions to return observables instead of promises.

**Comment by NetanelBasal** [comment]
What's our status here?

**Comment by zarghamkhandev** [comment]
> What's our status here?

`persisted-query` and `is-fetching` are left to be tested, will try to get this done asap.

File: packages/ng-query/src/lib/is-mutating.ts [hunk 6]
```diff
@@ -24,17 +24,20 @@ class IsMutating {
   ): Observable<number> {
     const [filters] = parseMutationFilterArgs(arg1, arg2);
 
-    return new Observable<number>((obs) =>
+    return new Observable<number>((obs) => {
+      obs.next(0);
```
Comment (zarghamkhandev) [inline packages/ng-query/src/lib/is-mutating.ts:28]: saw this in `vue-query` tests, shouldn't initial value be 0 ? @NetanelBasal

Comment (NetanelBasal) [reply packages/ng-query/src/lib/is-mutating.ts:28]: Yep

Comment (NetanelBasal) [reply packages/ng-query/src/lib/is-mutating.ts:28]: Hmmm, actually, not. Where did you see this?

Comment (zarghamkhandev) [reply packages/ng-query/src/lib/is-mutating.ts:28]: https://github.com/TanStack/query/blob/main/packages/vue-query/src/__tests__/useIsMutating.test.ts#L16

Comment (NetanelBasal) [reply packages/ng-query/src/lib/is-mutating.ts:28]: We should take the observer.getCurrentResult() and not hard-coded 0

Comment (zarghamkhandev) [reply packages/ng-query/src/lib/is-mutating.ts:28]: I believe it should be `obs.next(this.instance.isMutating(filters))` in this case.

**Comment by zarghamkhandev** [comment]
@NetanelBasal  the PR is ready for review. Some helper methods are left to be tested but they will come in separate PR.

File: packages/ng-query/src/lib/persisted-query.ts [hunk 7]
```diff
@@ -42,7 +42,7 @@ class PersistedQuery {
   >(
     queryObserverOptionsFn: (
       key: TQueryKey,
-      params: Params
+      params?: Params
```
Comment (zarghamkhandev) [inline packages/ng-query/src/lib/persisted-query.ts:45]: @NetanelBasal I believe params can be optional.

**Comment by NetanelBasal** [comment]
Great. Please update the [workflow](https://github.com/ngneat/query/blob/main/.github/workflows/ci.yml#L3) to run on new pull requests like [here](https://github.com/ngneat/until-destroy/blob/master/.github/workflows/until-destroy.yml#L3).

**Comment by zarghamkhandev** [comment]
✅

**Comment by NetanelBasal** [comment]
@zarghamkhandev tests failed https://github.com/ngneat/query/actions/runs/3359364249/jobs/5567311741

**Comment by zarghamkhandev** [comment]
fixed ✅

**Pull request merged**